sdk-core/src/main/java/dev/restate/sdk/core/RestateServerCall.java

```diff
@@ -161,11 +162,24 @@ private void pollInput() {
                         () -> {
                           Objects.requireNonNull(listener);
 
-                          // PollInput can only be result
-                          MessageLite message = deferredValue.toReadyResult().getResult();
-
-                          LOG.trace("Read input message:\n{}", message);
-                          listener.onMessageAndHalfClose(message);
+                          final ReadyResult<MessageLite> pollInputReadyResult =
+                              deferredValue.toReadyResult();
+
+                          if (pollInputReadyResult.isSuccess()) {
+                            final MessageLite message = pollInputReadyResult.getResult();
+                            LOG.trace("Read input message:\n{}", message);
+                            listener.onMessageAndHalfClose(message);
+                          } else {
+                            syscalls.writeOutput(
+                                pollInputReadyResult.getFailure(),
+                                SyscallCallback.ofVoid(
+                                    () ->
+                                        LOG.trace(
+                                            "Wrote output message:\n{}",
+                                            pollInputReadyResult.getFailure()),
+                                    this::onError));
+                            syscalls.close();
```

Comment:
I think invoking simply `this#close` with the terminal exception in the status should be fine here